src/util/object_gc_stats.rs

```diff
@@ -0,0 +1,159 @@
+                    let abs_gc_byte_offset = (object.to_address() - obj_lowest_addr) as isize
+                        + VM::VMObjectModel::GC_BYTE_OFFSET;
+                    // e.g. there is more than 8 bytes from lowest object address to gc byte
+                    if abs_gc_byte_offset >= constants::BYTES_IN_ADDRESS as isize {
+                        obj_lowest_addr // forwarding word at the lowest address of the object storage
+                    } else {
+                        obj_lowest_addr + constants::BYTES_IN_ADDRESS // forwarding word at the first word after the lowest address of the object storage
+                    }
+                } else {
+                    obj_lowest_addr // forwarding word at the lowest address of the object storage
+                }
+            }
+        }
+    }
+
+    #[cfg(target_endian = "big")]
+    fn get_object_status_word_address<VM: VMBinding>(object: ObjectReference) -> Address {
+        unimplemented!()
+    }
+
+    pub fn read<VM: VMBinding>(object: ObjectReference) -> usize {
+        unsafe {
+            Self::get_object_status_word_address::<VM>(object)
+                .atomic_load::<AtomicUsize>(Ordering::SeqCst)
+        }
+    }
+
+    pub fn write<VM: VMBinding>(object: ObjectReference, val: usize) {
+        trace!("GCForwardingWord::write({:#?}, {:x})\n", object, val);
+        unsafe {
+            Self::get_object_status_word_address::<VM>(object)
+                .atomic_store::<AtomicUsize>(val, Ordering::SeqCst)
+        }
+    }
+
+    pub fn compare_exchange<VM: VMBinding>(
+        object: ObjectReference,
+        old: usize,
+        new: usize,
+    ) -> bool {
+        // TODO(Javad): check whether this atomic operation is too strong
+        let res = unsafe {
+            Self::get_object_status_word_address::<VM>(object)
+                .compare_exchange::<AtomicUsize>(old, new, Ordering::SeqCst, Ordering::SeqCst)
+                .is_ok()
+        };
+        trace!(
+            "\nGCForwardingWord::compare_exchange({:#?}, old = {:x}, new = {:x}) -> {}\n",
+            object,
+            old,
+            new,
+            res
+        );
+        res
+    }
+}
+
+// (This function is only used internal to the `util` module)
+//
+// This function checks whether the forwarding word and GC byte can be unified (= the forwarding word fits in the word that contains the GC byte).
+//
+// Returns `None` if the forwarding word and GC byte can not be unified, or if unifiable,
+// returns `Some(fw_offset)`, where `fw_offset` is the offset of the forwarding word relative to `GC_BYTE_OFFSET`.
+//
+// A return value of `Some(fw_offset)` implies that GC byte and forwarding word can be loaded/stored with a single instruction.
+//
+pub(super) fn unifiable_gcbyte_forwarding_word_offset<VM: VMBinding>() -> Option<isize> {
+    let gcbyte_dealignment = VM::VMObjectModel::GC_BYTE_OFFSET % constants::BYTES_IN_WORD as isize;
```

Comment:
`dealignment` may be better renamed to `gcbyte_lshift` as it effectively represents the number of left shifts to perform on the gcbyte value before it can be stored to the same word as `forwarding word`.
In jikesrvm, it is 0 because the gcbyte can simply be `or`ed with the forwarding word.
In openjdk-64, it is 7 because the gcbyte should be left shifted 7 times before being `or`ed with the `forwarding word`.